src/main/java/uk/co/fivium/dmda/emailmessages/DatabaseMessageStorer.java

```diff
@@ -102,26 +101,29 @@ private void setHeaderBindIfExists(String pStoreQuery, OraclePreparedStatement p
   private void storeMessageBody(EmailMessage pEmailMessage, String lDatabaseName, OraclePreparedStatement pStatement, String pStoreQuery)
   throws SQLException, ParserConfigurationException {
     for (EmailRecipient lRecipient : pEmailMessage.getRecipients()){
-      String lRecipientDatabase = SMTPConfig.getInstance().getDatabaseForRecipient(lRecipient.mDomain);
-      if(lDatabaseName.equals(lRecipientDatabase)){
-        // Bind values to the bind variables if they exist in the store query.
-        setStringAtNameIfExists(pStoreQuery, pStatement, BindParams.MAIL_ID.getText(), pEmailMessage.getMailId());
-        setStringAtNameIfExists(pStoreQuery, pStatement, BindParams.USER.getText(), lRecipient.mUser);
-        setStringAtNameIfExists(pStoreQuery, pStatement, BindParams.FROM.getText(), pEmailMessage.getFrom());
-        setStringAtNameIfExists(pStoreQuery, pStatement, BindParams.RECIPIENT.getText(), lRecipient.mEmailAddress);
-        setStringAtNameIfExists(pStoreQuery, pStatement, BindParams.REMOTE_HOSTNAME.getText(), pEmailMessage.getRemoteHostname());
-        setStringAtNameIfExists(pStoreQuery, pStatement, BindParams.REMOTE_ADDRESS.getText(), pEmailMessage.getRemoteAddress());
-        setBlobAtNameIfExists(pStoreQuery, pStatement, BindParams.MESSAGE_BODY.getText(), pEmailMessage.getDataStream());
-        setStringAtNameIfExists(pStoreQuery, pStatement, BindParams.SUBJECT.getText(), pEmailMessage.getSubject());
-        setDateAtNameIfExists(pStoreQuery, pStatement, BindParams.SENT_DATE.getText(), pEmailMessage.getSentDate());
-        setHeaderBindIfExists(pStoreQuery, pStatement, pEmailMessage);
-
-        pStatement.execute();
+      for (String recipientDatabase : SMTPConfig.getInstance().getDatabasesForRecipient(lRecipient.mDomain)) {
```

Comment:
Is there a reason this re-queries the list of DBs for that recipient?
It seems to already get the DBs for the given recipient in this methods caller `storeMessage` (via `getUniqueDatabaseDestinations`), and so `lDatabaseName` is already guaranteed to be in the list of recipient databases.

I suppose a belt and braces check doesn't hurt and the former removes duplicates and here it doesn't. But it seems almost undesirable that if you have multiple matching rules for the same DB that you would get duplicate identical rows? 
Not a big deal either way so happy to leave it as is if you prefer.

Reply:
As discussed this is doing a method call per database, checking if each recipient is implicated in this database. I'll change it to a contains match so it's more clear